cache/fast_lru_cache.h

```diff
@@ -117,42 +141,68 @@ class LRUHandleTable {
+  // Returns a pointer to the inserted handle. The argument old is
+  // an output argument, which is a reference to a pointer to an
+  // existing handle that matches the key/hash, or nullptr if no such
+  // handle exists.
+  LRUHandle* Insert(LRUHandle* h, LRUHandle** old);
+  // Returns a pointer to the handle matching the key/hash, or nullptr
+  // if not present. Assumes the element is in the table.
   LRUHandle* Remove(const Slice& key, uint32_t hash);
 
+  void Exclude(LRUHandle* e);
+
+  void Assign(int slot, LRUHandle* src);
+
   template <typename T>
   void ApplyToEntriesRange(T func, uint32_t index_begin, uint32_t index_end) {
     for (uint32_t i = index_begin; i < index_end; i++) {
-      LRUHandle* h = list_[i];
-      while (h != nullptr) {
-        auto n = h->next_hash;
-        assert(h->InCache());
+      LRUHandle* h = &array_[i];
+      if (h->IsPresent()) {
         func(h);
-        h = n;
       }
     }
   }
 
   int GetLengthBits() const { return length_bits_; }
 
-  // Return the address of the head of the chain in the bucket given
-  // by the hash.
-  inline LRUHandle** Head(uint32_t hash);
+  uint32_t GetOccupancy() const { return occupancy_; }
 
  private:
-  // Return a pointer to slot that points to a cache entry that
-  // matches key/hash.  If there is no such cache entry, return a
-  // pointer to the trailing slot in the corresponding linked list.
-  LRUHandle** FindPointer(const Slice& key, uint32_t hash);
+  int FindElement(const Slice& key, uint32_t hash, int& probe,
+                  int displacement);
+
+  int FindPresentElement(const Slice& key, uint32_t hash, int& probe,
+                         int displacement);
+
+  int FindAvailableSlot(const Slice& key, int& probe, int displacement);
+
+  int FindPresentElementOrAvailableSlot(const Slice& key, uint32_t hash,
+                                        int& probe, int displacement);
+
+  // Returns the index of the first probe that contains a handle e
+  // such that cond(e) is true.  If the probe sequence ends without
+  // a match, returns a pointer to the first slot available in the
+  // sequence.  Otherwise, if no match is found and no slot is
+  // available, returns -1. For every handle e probed, except any
+  // matching slot, we update e->displacements += displacement.
+  // The variable probe is updated such that consecutive calls to
+  // FindSlot use every probe number, without repetition.
+  int FindSlot(const Slice& key, std::function<bool(LRUHandle*)> cond,
+               int& probe, int displacement);
 
   // Number of hash bits (upper because lower bits used for sharding)
   // used for table index. Length == 1 << length_bits_
   int length_bits_;
 
-  // The table consists of an array of buckets where each bucket is
-  // a linked list of cache entries that hash into the bucket.
-  std::unique_ptr<LRUHandle*[]> list_;
+  uint32_t occupancy_;
+
+  // TODO(guido) Cache alignment.
+  // TODO(guido) unique_ptr?
```

Comment:
I think you ideally want `std::unique_ptr<LRUHandle[]>`. Unique_ptr-to-array is a little tricky